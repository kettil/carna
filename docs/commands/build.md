# The `build` task

Build the application

## Options

| Options | Description       |
| ------- | ----------------- |
| --ci    | Run it in CI mode |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
