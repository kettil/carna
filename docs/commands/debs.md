# The `debs` task

Checks if there are orphaned packages

## Options

| Options | Description       |
| ------- | ----------------- |
| --ci    | Run it in CI mode |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
