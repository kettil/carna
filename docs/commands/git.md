# The `git` tasks

The tasks will be automatically called by husky during commit.

## The `git commit` task

Checks the files in staged with type, prettier and eslint.

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |

## The `git msg` task

Checks the git commit message for a consistent structure.

### Options

| Options | Description                     |
| ------- | ------------------------------- |
| --edit  | Path to the COMMIT_EDITMSG file |

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |

## The `git push` task

Checks the files with type, prettier and eslint.

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
