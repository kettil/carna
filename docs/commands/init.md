# The `init` task

Creates a new project with all config files in the current folder.

## Options

| Options    | short | Description                                                  |
| ---------- | ----- | ------------------------------------------------------------ |
| --package  | -p    | Project is created as a package                              |
| --cli      | -c    | Extends the package with CLI features (required `--package`) |
| --noCommit |       | No initial commit is executed at the end                     |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
