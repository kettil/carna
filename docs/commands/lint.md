# The `lint` task

Run the code quality tools

## Options

| Options | Description                                      |
| ------- | ------------------------------------------------ |
| --ci    | Run it in CI mode                                |
| --only  | Run a single code quality tool (required `--ci`) |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
