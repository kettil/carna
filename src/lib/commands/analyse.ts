import { createBuilder, createHandler } from '../cli/yargs';
import type { AnalyseProps } from '../tasks/analyseTask';
import { analyseServices, analyseTask } from '../tasks/analyseTask';

const command = 'analyse';
const desc = 'Run code quality tools';

const aliases = ['$0'];
const options = { group: `${command}-Options` } as const;
const boolOptions = { ...options, type: 'boolean', default: false } as const;

const handler = createHandler<AnalyseProps>(analyseTask);
const builder = createBuilder<AnalyseProps>(command, (yargs) =>
  yargs.options({
    all: {
      ...boolOptions,
      alias: 'a',
      describe: 'All files are checked and not only the changed/new files',
    },
    only: {
      ...options,
      alias: 'o',
      choices: analyseServices,
      describe: 'Run a single code quality tool',
    },
  }),
);

export { aliases, command, desc, builder, handler };
