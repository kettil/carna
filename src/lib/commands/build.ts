import { createBuilder, createHandler } from '../cli/yargs';
import { buildTask, BuildProps } from '../tasks/buildTask';

const command = 'build';
const desc = 'Run the build process';

const options = { group: `${command}-Options` } as const;
const boolOptions = { ...options, type: 'boolean', default: false } as const;

const handler = createHandler<BuildProps>(buildTask);
const builder = createBuilder<BuildProps>(command, (yargs) =>
  yargs.options({
    watch: {
      ...boolOptions,
      alias: 'w',
      describe: 'Watch files for changes and rebuild the changed files (without type files)',
    },
  }),
);

export { command, desc, builder, handler };
