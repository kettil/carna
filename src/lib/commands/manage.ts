import { createBuilder, createHandler } from '../cli/yargs';
import type { ManageProps } from '../tasks/manageTask';
import { manageTask } from '../tasks/manageTask';

const command = 'manage';
const desc = 'Run general conditions check';

const handler = createHandler<ManageProps>(manageTask);
const builder = createBuilder<ManageProps>(command, (yargs) => yargs);

export { command, desc, builder, handler };
