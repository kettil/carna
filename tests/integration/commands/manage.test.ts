import * as fs from 'fs';
import type { Argv } from 'yargs';
import { handler, builder, command, desc } from '../../../src/lib/commands/manage';
import type { PropsGlobal } from '../../../src/lib/types';
import { getArgv } from '../../shared/setup/argv';
import { getReadFileWithHooksFiles } from '../../shared/setup/readFileFiles';

describe('command manage', () => {
  test('it should be complete the yargs command structure', () => {
    expect(typeof command).toBe('string');
    expect(typeof desc).toBe('string');
    expect(typeof builder).toBe('function');
    expect(typeof handler).toBe('function');
  });

  test('it should be create the expected builder', () => {
    expect.assertions(1);

    const yargs = {
      options: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('options');

        return yargs;
      }),
      usage: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('usage');

        return yargs;
      }),

      help: jest.fn(() => yargs),
      version: jest.fn(() => yargs),
      showHelpOnFail: jest.fn(() => yargs),
    } as unknown as Argv<PropsGlobal>;

    builder(yargs);
  });

  test('it should work', async () => {
    await handler(getArgv({}));

    expect(true).toBeTruthy();
  });

  test('it should work with hooks', async () => {
    (fs as any).setMockReadFileFiles(getReadFileWithHooksFiles());

    await handler(getArgv({}));

    expect(true).toBeTruthy();
  });
});
