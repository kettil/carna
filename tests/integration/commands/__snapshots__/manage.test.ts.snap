// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command manage it should be create the expected builder: usage 1`] = `
Array [
  "Usage: $0 manage [options]",
]
`;

exports[`command manage it should work with hooks: depcheck 1`] = `
Array [
  "/path/to/project",
  Object {
    "detectors": Array [
      [Function],
      [Function],
    ],
    "ignoreMatches": Array [
      "@types/node",
      "typescript",
      "carna",
    ],
    "ignorePatterns": Array [
      "node_modules",
      "build",
    ],
    "parsers": Object {
      "**/*.js": [Function],
      "**/*.jsx": [Function],
      "**/*.ts": [Function],
      "**/*.tsx": [Function],
    },
    "specials": Array [
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
    ],
  },
]
`;

exports[`command manage it should work: depcheck 1`] = `
Array [
  "/path/to/project",
  Object {
    "detectors": Array [
      [Function],
      [Function],
    ],
    "ignoreMatches": Array [
      "@types/node",
      "typescript",
      "carna",
    ],
    "ignorePatterns": Array [
      "node_modules",
      "build",
    ],
    "parsers": Object {
      "**/*.js": [Function],
      "**/*.jsx": [Function],
      "**/*.ts": [Function],
      "**/*.tsx": [Function],
    },
    "specials": Array [
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
    ],
  },
]
`;
